### PR TITLE
Feature/guard for json

### PR DESCRIPTION
**Pull request by pontusBerg**
## Changes

`gulpfile.js`

  `stacGenerateIndex()`
  - Added check for every dataset to make sure that it can be stringified. 
  - Added stringify check for all datasets combined that goes in the index stac file.
  
  
  `stacGenerate()`
  
  - Added parse check for every dataset 
  
  
`tests`

  `stac-json.test.js`
  
  - Setup test which checks if every file in the `/_output/stac` folder can be parsed as JSON, and if the parsed file is an object.  These tests are run automatically on commit to ensure that the checks in the `gulpfile` doesn't miss a case. 
  

## Other notes 

I acidentally ran prettier on the gulpfile. Let me know if this should be reverted.

### CODE REVIEW

File: stac-json.test.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+require('jest')
+const fs = require('fs');
+const path = require('path');
+
+const jsonDir = './_output/stac/';
+const files = fs.readdirSync(jsonDir);
```
Comment (dgostencnik) [inline stac-json.test.js:6]: This test crashes if output folder doesn't exists. This use case ( user wants to commit something without building firs) should probably be handled.

File: package.json [hunk 2]
```diff
@@ -21,6 +22,7 @@
     "gulp-rename": "^1.4.0",
     "gulp-yaml": "^2.0.4",
     "handlebars": "^4.3.0",
+    "husky": "^7.0.4",
```
Comment (dgostencnik) [inline package.json:25]: This doesn't really work if you don't install husky manually. In order to get this working, you need to include prepare/postInstall script in  `scripts` section of package.json

https://blog.typicode.com/husky-git-hooks-autoinstall/

```
 "scripts": {
    "postinstall": "husky install"
  }
```

File: stac-json.test.js (hunk 1 above)
Comment (pontusBerg) [reply stac-json.test.js]: Tests are now skipped if no _output/stac folder exists.

File: package.json (hunk 2 above)
Comment (pontusBerg) [reply package.json:26]: Added postinstall script

File: stac-json.test.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+if (isExistingDirectory(jsonDir)) {
+    testif = test
+    files = fs.readdirSync(jsonDir);
+} else {
+    testif = test.skip
+    files = []
+}
+
+files.forEach((file) => {
+    testif(`Should parse ${file} in STAC folder as JSON.`, () => {
```
Comment (dgostencnik) [inline stac-json.test.js:19]: I don't think it is neccessary to have this `testif` variable. fftr

**Review by dgostencnik — APPROVED** [review]
It would probably be even better to set up workflow and execute test for each push action on github. 

https://docs.github.com/actions/automating-builds-and-tests/building-and-testing-nodejs-or-python?langId=nodejs

**Comment by pontusBerg** [comment]
> It would probably be even better to set up workflow and execute test for each push action on github.

I am not a 100% sure how I would do this. The `_output` folder is gitignored, and it seems like the stac jsons are taken care of in the `buildspec.yaml` file. Or am I missing something?

**Comment by dgostencnik** [comment]
> > It would probably be even better to set up workflow and execute test for each push action on github.
> 
> I am not a 100% sure how I would do this. The `_output` folder is gitignored, and it seems like the stac jsons are taken care of in the `buildspec.yaml` file. Or am I missing something?

You can do this literally by copying example from link above. It will run `build` script, which will create _output folder, and  afterwards `test` script.  

If you want to play with github actions, `act` is a nice tool to test them locally.

However, I don't think it is really neccessary as there are already two checks for valid stac jsons (on build by gulp.js and on commit by husky). 

I've already approved this PR, but I think it is ignored as I don't have write privileges on this repo.

**Review by dthiex — APPROVED** [review]

**Comment by dthiex** [comment]
> I've already approved this PR, but I think it is ignored as I don't have write privileges on this repo.

@dgostencnik added you to this repo for the future

@pontusBerg Approved the MR already, so if everything is resolved you can merge this.

**Pull request merged**